### PR TITLE
Support creating config files from palette modules

### PR DESCRIPTION
**Pull request by jan-warchol**
- expand convert.py to handle conversion from many formats and into multiple color spaces
- let colorfiller load palettes from special modules, process them with convert.py and fill templates using that

See individual commits for details, and read `utils/README.md` for examples showing how to test the code.

@mcFrax can you review please?

### CODE REVIEW

File: utils/colorfiller.py [hunk 1]
```diff
@@ -3,66 +3,70 @@
...
 import re
+import convert
+import importlib
 
-tmpl_ext = '.template'
+USAGE = """Evaluate templates using specified palette.
 
-marker_re = re.compile(r'!!COL(?P<delim>.)(?P<format>.*?)(?P=delim)')
+Takes two arguments - path to a template file or to a folder that will be
+recursively searched for templates and name of the palette module to use."""
```
Comment (mcFrax) [inline utils/colorfiller.py]: Possibly a bit more flexible design could be to take palette as first argument and then any number of path arguments, or to use option (`--palette`). Well, this is not actually a review comment, its more like a feature suggestion.

File: utils/convert.py [hunk 2]
```diff
@@ -1,52 +1,145 @@
...
-    else:
-        return convert_color(color, sRGBColor)
+    return parsed, base
+
+
+class Color(object):
+    # some of the bright versions of colors are slightly out of RGB gamut -
+    # it's not a big deal, we just clamp them.
+    @staticmethod
+    def clamp(color):
```
Comment (mcFrax) [inline utils/convert.py]: Why is it clamping only from one end? I mean, maybe there is no need to clamp negative values, but that is a bit unexpected for `clamp` function. Rename?

Yep, I'm a bit pesky, but I suppose that is what you expected from me :P

File: utils/convert.py [hunk 3]
```diff
@@ -83,13 +85,13 @@ def __init__(self, spec):
...
+    def __str__(self):
+        return "   ".join(["{}"]*4).format(
+            self.lab,
+            self.hsv,
+            self.srgb,
+            self.apple,
+        )
 
     def __repr__(self):
         return "\n".join(["{}"]*4).format(
```
Comment (mcFrax) [inline utils/convert.py]: Err... why?

That doesn't look like a good __repr__ at all. At least wrap it in some `<Color object ... >` or something. Don't risk anything like `AssertionError: [1] != [1]`.

File: utils/convert.py [hunk 4]
```diff
@@ -1,60 +1,116 @@
...
-    coords = hex_color_re.match(color)
+# def rgb_str(self):
+#     return "{:>3.0f},{:>3.0f},{:>3.0f}".format(
+#         self.rgb_r*255, self.rgb_g*255, self.rgb_b*255
+#     )
+
+LabColor.__str__      = lab_str
+HSVColor.__str__      = hsv_str
+sRGBColor.__str__     = sRGBColor.get_rgb_hex
+AppleRGBColor.__str__ = AppleRGBColor.get_rgb_hex
```
Comment (mcFrax) [inline utils/convert.py]: Ouch, how ugly (modyfing classes on the fly). Ouch, how pythonic (modyfing classes on the fly).

About formatting: I'm rather against such alignment of assignments, it ruins statements own readability for questionable esthetic reasons.

**Comment by mcFrax** [comment]
I left some comments on code, but generally lgtm.

File: utils/convert.py (hunk 2 above)
Comment (jan-warchol) [inline utils/convert.py]: `colormath` library clamps negative values on its own.

File: utils/convert.py (hunk 4 above)
Comment (jan-warchol) [inline utils/convert.py]: Hehe :)
As for formatting, opinions vary; personally I prefer aligning assignments.

**Comment by jan-warchol** [comment]
Addressed all comments. Will merge in a few minutes.

File: utils/convert.py (hunk 3 above)
Comment (jan-warchol) [inline utils/convert.py]: Changed. I hope it's acceptable now.

**Pull request merged**